lib/ansible/plugins/action/__init__.py

```diff
@@ -841,13 +841,9 @@ def _update_module_args(self, module_name, module_args, task_vars):
         # make sure modules are aware if they need to keep the remote files
         module_args['_ansible_keep_remote_files'] = C.DEFAULT_KEEP_REMOTE_FILES
 
-        # make sure all commands use the designated temporary directory if created
-        if self._is_become_unprivileged():  # force fallback on remote_tmp as user cannot normally write to dir
-            module_args['_ansible_tmpdir'] = None
-        else:
-            module_args['_ansible_tmpdir'] = self._connection._shell.tmpdir
-
-        # make sure the remote_tmp value is sent through in case modules needs to create their own
+        # We used to have this set to self._connection._shell.tmpdir but to keep things simple we don't anymore
+        # always have a module use the remote_tmp set by the shell plugin and have the module create their own.
+        module_args['_ansible_tmpdir'] = None  # TODO: Remove this after we know it doesn't break anything
```

Comment:
Several modules rely on this, but they won't error out as they are using `mksdtemp(dir=module.tmpdir)` based on this value, when `None` it falls back to `$TEMP, $TMP, /tmp` discovery .. also now modules that were not cleaning up their own temps will now be leaving them behind.

Reply:
This proves my point that the module side tempdirs are complicated enough that how they actually work today isn't well understood. `module.tmpdir` is all handled by the [tmpdir](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L687) property and this is the logic as it stands

* If `self._tmpdir` is set, return that
* If not create a module managed tempdir
    * Expand user/vars of `self._remote_tmp`
    * Create `self._remote_tmp` with `0o700` if it does not exist, always warns the user
    * Create a tempdir in `self._remote_tmp` with `tempfile.mkdtemp`
    * Register `shutil.rmtree` to run atexit to delete the tempdir
    * Sets `self._tmpdir` to this new value so future calls return that

The `_ansible_tmpdir` variable passed from the controller is set to `self._tmpdir` so any module that calls `module.tmpdir` will use the controller created one. The `_ansible_remote_tmp` is set to `self._remote_tmp` and is used by `module.tmpdir` to create the module scoped tempdir if `_ansible_tmpdir` is not set. The only thing this change does is to stop sending `_ansible_tmpdir` so that  `self._tmpdir` is `None` and any module that calls `module.tmpdir` will just have a module managed tempdir under `_ansible_remote_tmp`. The only thing I can see this breaking is if a module called `module._tmpdir` which is already problematic for 3 reasons

1. It's not always set, so they could be getting `None` for a temp dir instead of the user desired one whereas `remote_tmp` is always set
2. Using an underscore prefixed attribute like `module._tmpdir` implies that it is not for public use, and I agree modules should be using `module.tmpdir`
3. Based on the fallback logic for underprivileged become, this *COULD* be readable by other users whereas `remote_tmp` is by default scoped to the user's home at `~/.ansible/tmp`. A user would have to explicitly set `remote_tmp` to a bad location for this to happen where underprivilege system_tmp just happens by default

So while these are breaks in the sense the behaviour changes they are behavioural changes to simplify how this all works and the only modules that are affected aren't doing this right, and even so they would only work somethings, i.e. when `_ansible_tmpdir` is actually set (which is not guaranteed).

So by only using `_ansible_remote_tmp` this is what we gain

1. We always ensure the module tempdir is set to whatever the shell's `remote_tmp` is set to, not the mixture of it could be `remote_tmp`, could be one of `system_tmpdirs`, it's just `remote_tmp`
2. Because of the above the tempdir will be in the user's homedir `~/.ansible/tmp` by default and an Ansible user would have to explicitly set it to something else to have it be outside that area
3. If for some reason a module is doing something wrong and accessed `module._tmpdir` directly it forces their hand to actually change it to `module.tmpdir`. Right now it only sometimes is set whereas in the future it will always be `None` until` module.tmpdir` is called for the first time

Reply:
I don't see how this is a simplification, we just give modules the information about the dir the controller would have used, but it doesn't force the module to do anything. While this change will alter the behaviour of existing modules that rely on it.

Reply:
It simplifies it in the fact that when talking about a remote tmpdir for a module it's only ever controlled by 1 thing and 1 thing only. Right now it can be

* One created by the controller, this can be
    * `remote_tmp` for normal users
    * `system_tempdirs` if become to a non-admin user is used
    * Nothing at all because the controller didn't need to create one (pipelining)
* The `remote_tmp` if the above was not created

Try explaining this to an end user and their eyes will be glazed over. By taking away the controller tempdir as an option for a module you are left with 1 config option that is always used when a module uses a tempdir.

I don't see how this alters a behaviour that a module can rely on. If a module was relying on `module.tmpdir` being set to `_ansible_tmpdir` it would then be broken when pipelining is used, or become to an underprivileged user happens as `_ansible_tmpdir` won't be set at all. It just falls back to the same behaviour I'm proposing here to be the behaviour that always happens.

This won't solve controller managed tempdirs as that needs to handle a lot more complex situations but we can at least simplify module managed tempdirs this way.

Reply:
that still happens no matter what info you pass to the module ... unsure how removing this info for the module to CHOOSE to share the dir is a simplification. 

Speaking of the end user, we CAN show them the tempdir being used since that info is available on the controller, but not when the module itself generates it (and many do).

Reply:
> that still happens no matter what info you pass to the module

It does if the controller needs to create a tempdir but typically that's for things like copy where the controller needs to copy a temp file across or when pipelining is not enabled. This is not something that happens all the time so already we have a branching logic for how a module uses a tempdir. I'm trying to remove this logic and just always state that a module always uses it's own tempdir at `remote_tmp`.

> unsure how removing this info for the module to CHOOSE to share the dir is a simplification.

It's not a simplification of the code as that mostly stays the same but a simplification of how tempdirs used in a module work. With this change we can simply say that tempdirs used by a module is based on `remote_tmp`. This isn't all tempdirs but just ones created by the module.

Reply:
The thing is that you are removing an existing feature, not simplifying anything, nothing FORCES modules to use this, many don't, that is why we ALSO supply remote_tmp (though that value is not always usable depending on become and other circumstances). Many modules don't use either value and opt to handle it completely on their own.

But those that do use it, rely on this behaviour and making this change won't 'break' things, it will start leaving dangling temp files or attempt to create them in an unexpected directory.

Reply:
> though that value is not always usable depending on become and other circumstances

The only case when that is not usable is when someone set's it to a non-default value that uses an absolute path. The same problem would exist if the path was edited and the connection user did not have access to that location. At least with the default of `~/.ansible/tmp` will have each module tmpdir actually located inside the executing user's home direction and not what we have today which is potentially in a globally readable/writable location like `/tmp`.

> Many modules don't use either value and opt to handle it completely on their own.

If they are doing that they are wrong today. Say they are using `_ansible_tmpdir` directly and ignoring `_ansible_remote_tmp` then they are only setting the tmpdir in some cases and not all. That silent break you are worried about is happening today. These modules should be updating their code to use the proper mechanism we expose which is `module.tmpdir`.

> But those that do use it, rely on this behaviour and making this change won't 'break' things in a way that alerts anyone,

Once again, if they are relying on this behaviour they are getting silently failures today as `_ansible_tmpdir` is not always set to a value. So instead of it sometimes works but only in x, y, z, it's now it never works because my code is wrong and I shouldn't be relying on these vars.

Ultimately you don't agree and I'm not going to push this any further. I think it's useful to try and simplify all this as best we can and even think this improves the security with module side tempdirs as it's no longer using the potentially world readable/writable ones the controller may have to generate.

Reply:
> The only case when that is not usable is when someone set's it to a non-default value that uses an absolute path. 
No, there are other ways, for example, since different privilege escalation setups can translate `$HOME` and `~/` in different ways so sometimes (we hope most times) it resolves to the final user, but it can be configured to use the data from the login user.

>If they are doing that they are wrong today. Say they are using _ansible_tmpdir directly and ignoring _ansible_remote_tmp
In many cases they don't use either, but I used those to point out it is currently a CHOICE module authors have made, it is not something we forced. I'm against removing a option we provided, if you want to change the modules to use a different directory or logic, that can be done at module level, but changing a provided feature of the controller that they opted into is not the way. We already provide the option you want in the form of the other variable.

> Once again, if they are relying on this behaviour they are getting silently failures today as _ansible_tmpdir is not always set to a value. 
And this is a bug, it should not return `None` on that condition as I stated above, since it should fallback on `system_tempdirs` which are also configurable by the user.